src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/prisonersearch/PrisonerSearchService.kt

```diff
@@ -13,24 +14,40 @@ private val log = KotlinLogging.logger {}
 class PrisonerSearchService(private val prisonerSearchClient: PrisonerSearchClient) {
   fun lookupPrisoner(barcodeRecipient: BarcodeRecipient) {
     val prisonerSearchRequest = prisonerSearchRequest(barcodeRecipient)
-    try {
-      val matchPrisonersResponse = prisonerSearchClient.matchPrisoners(prisonerSearchRequest)
-      log.info { "Found ${matchPrisonersResponse.matches.size} potential prisoner matches" }
-    } catch (webClientResponseException: WebClientResponseException) {
-      if (webClientResponseException.statusCode == HttpStatus.FORBIDDEN) {
-        log.info { "Client token does not have correct role to call prisoner-search" }
-      } else {
-        log.error { "Failed to call prisoner-search - statusCode ${webClientResponseException.statusCode}" }
+    prisonerSearchClient.matchPrisoners(prisonerSearchRequest)
+      .onErrorResume {
+        with(it) {
+          if (this is WebClientResponseException) {
+            if (statusCode == HttpStatus.FORBIDDEN) {
+              log.info { "Client token does not have correct role to call prisoner-search" }
+            } else {
+              log.error { "Failed to call prisoner-search - statusCode $statusCode" }
+            }
+          } else {
+            log.error("Failed to call prisoner-search", it)
+          }
+        }
+          .let { Mono.empty() }
```

Comment:
Looks like there is some kotlin language support for this: https://github.com/ministryofjustice/hmpps-incentives-api/blob/main/src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/PrisonApiService.kt#L27

Reply:
Looks interesting, but also looks like it needs a bit more thought in respect our project.
Their catch block is on the webclient iteself, rather than in the service - I _think_ for what we are trying to do our error handling probably wants to be at the service level?
They also appear to be using coroutines which I know very little about (but they sound cool 👍 ) and I think they've brought in a library in their gradle file to support it (eg: none of these are currently on our classpath: 
```
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.emptyFlow
```
)